server/resp/src/main/java/org/infinispan/server/resp/commands/cluster/SHARDS.java

```diff
@@ -0,0 +1,255 @@
+package org.infinispan.server.resp.commands.cluster;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.commons.util.concurrent.CompletableFutures;
+import org.infinispan.configuration.cache.Configuration;
+import org.infinispan.distribution.DistributionManager;
+import org.infinispan.distribution.ch.ConsistentHash;
+import org.infinispan.manager.CacheManagerInfo;
+import org.infinispan.manager.ClusterExecutor;
+import org.infinispan.manager.EmbeddedCacheManager;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.remoting.transport.Transport;
+import org.infinispan.remoting.transport.jgroups.JGroupsAddress;
+import org.infinispan.security.actions.SecurityActions;
+import org.infinispan.server.resp.ByteBufferUtils;
+import org.infinispan.server.resp.Resp3Handler;
+import org.infinispan.server.resp.RespCommand;
+import org.infinispan.server.resp.RespRequestHandler;
+import org.infinispan.server.resp.cluster.NodeInformation;
+import org.infinispan.server.resp.commands.Resp3Command;
+import org.infinispan.topology.CacheTopology;
+import org.jgroups.stack.IpAddress;
+
+import io.netty.channel.ChannelHandlerContext;
+
+/**
+ * `CLUSTER SHARDS` command.
+ * <p>
+ * Use the {@link CacheTopology} and retrieves information based on the {@link ConsistentHash}. We broadcast the
+ * command to the current topology members to retrieve specific data from the nodes.
+ *
+ * @link <a href="https://redis.io/commands/cluster-shards/">CLUSTER SHARDS</a>
+ * @since 15.0
+ */
+public class SHARDS extends RespCommand implements Resp3Command {
+   private static final CompletionStage<RespRequestHandler> NO_CONSISTENT_HASH =
+         CompletableFuture.failedFuture(new IllegalStateException("No consistent hash available"));
+
+   public SHARDS() {
+      super(2, 0, 0, 0);
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, ChannelHandlerContext ctx, List<byte[]> arguments) {
+      AdvancedCache<?, ?> respCache = handler.cache();
+      DistributionManager dm = respCache.getDistributionManager();
+      if (dm == null) {
+         Configuration configuration = SecurityActions.getCacheConfiguration(respCache);
```

Comment:
The redis-server returns "(error) ERR This instance has cluster support disabled" when no clustering is enabled.